Source/ASVideoPlayerNode.mm

```diff
@@ -95,122 +96,128 @@ @implementation ASVideoPlayerNode
+  if (!(self = [self initWithAsset:asset])) {
     return nil;
   }
-
-  _asset = asset;
-  _videoComposition = videoComposition;
-  _audioMix = audioMix;
-  _loadAssetWhenNodeBecomesVisible = YES;
-
-  [self _init];
+  
+  _videoNode.videoComposition = videoComposition;
+  _videoNode.audioMix = audioMix;
 
   return self;
 }
 
-- (instancetype)initWithUrl:(NSURL *)url loadAssetWhenNodeBecomesVisible:(BOOL)loadAssetWhenNodeBecomesVisible
+- (void)_initControlsAndVideoNode
 {
-  if (!(self = [super init])) {
-    return nil;
-  }
+  _defaultControlsColor = [UIColor whiteColor];
+  _cachedControls = [[NSMutableDictionary alloc] init];
+  
+  _videoNode = [[ASVideoNode alloc] init];
+  _videoNode.delegate = self;
+  [self addSubnode:_videoNode];
+}
 
-  _url = url;
-  _asset = [AVAsset assetWithURL:_url];
-  _loadAssetWhenNodeBecomesVisible = loadAssetWhenNodeBecomesVisible;
+#pragma mark Deprecated
 
-  [self _init];
+- (instancetype)initWithUrl:(NSURL *)url
+{
+  return [self initWithURL:url];
+}
 
-  return self;
+- (instancetype)initWithUrl:(NSURL *)url loadAssetWhenNodeBecomesVisible:(BOOL)loadAssetWhenNodeBecomesVisible
+{
+  return [self initWithURL:url];
 }
 
 - (instancetype)initWithAsset:(AVAsset *)asset loadAssetWhenNodeBecomesVisible:(BOOL)loadAssetWhenNodeBecomesVisible
 {
-  if (!(self = [super init])) {
-    return nil;
-  }
+  return [self initWithAsset:asset];
+}
 
-  _asset = asset;
-  _loadAssetWhenNodeBecomesVisible = loadAssetWhenNodeBecomesVisible;
+- (instancetype)initWithAsset:(AVAsset *)asset videoComposition:(AVVideoComposition *)videoComposition audioMix:(AVAudioMix *)audioMix loadAssetWhenNodeBecomesVisible:(BOOL)loadAssetWhenNodeBecomesVisible
+{
+  return [self initWithAsset:asset videoComposition:videoComposition audioMix:audioMix];
+}
 
-  [self _init];
+#pragma mark - Setter / Getter
 
-  return self;
+- (void)setAssetURL:(NSURL *)assetURL
+{
+  // Call setter to 
+  self.asset = [AVAsset assetWithURL:assetURL];
 }
 
--(instancetype)initWithAsset:(AVAsset *)asset videoComposition:(AVVideoComposition *)videoComposition audioMix:(AVAudioMix *)audioMix loadAssetWhenNodeBecomesVisible:(BOOL)loadAssetWhenNodeBecomesVisible
+- (NSURL *)assetURL
 {
-  if (!(self = [super init])) {
-    return nil;
+  ASDN::MutexLocker l(__instanceLock__);
+  
+  if ([_pendingAsset isKindOfClass:AVURLAsset.class]) {
+    return ((AVURLAsset *)_pendingAsset).URL;
+  } else {
+    return _videoNode.assetURL;
   }
+}
 
-  _asset = asset;
-  _videoComposition = videoComposition;
-  _audioMix = audioMix;
-  _loadAssetWhenNodeBecomesVisible = loadAssetWhenNodeBecomesVisible;
-
-  [self _init];
+- (void)setAsset:(AVAsset *)asset
+{
+  ASDisplayNodeAssertMainThread();
 
-  return self;
+  ASDN::MutexLocker l(__instanceLock__);
+  
+  // Clean out pending asset
+  _pendingAsset = nil;
+  
+  // Set asset based on interface state
+  if ((ASInterfaceStateIncludesPreload(self.interfaceState))) {
+   _videoNode.asset = asset;
```

Comment:
Also, it maybe a good idea to avoid holding the instance lock while calling the video node. Doing so might potentially cause deadlocks if videoNode (or a subclass) for some reasons walks up the node hierarchy.